File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    dx = [1, -1, 0, 0]
    dy = [0, 0, -1, 1]
    
    # 1일 때는 그냥 값 넣고 1이 아니라면 최솟값 집어넣기
    q = deque()
    q.append((0,0))
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]) and maps[nx][ny] == 1:
                maps[nx][ny] += maps[x][y]
                q.append((nx, ny))
        # for x in maps:
        #     print(x)
        # print('--------------------')

    return a if (a:=maps[-1][-1]) > 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
import java.awt.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        int col = maps.length;
        int row = maps[0].length;
        
        int[][] cost = new int[col][row];
        for (int i = 0; i < col; i++) {
            for (int j = 0; j < row; j++) {
                cost[i][j] = Integer.MAX_VALUE;
            }
        }
        
        Deque<Point> q = new ArrayDeque<>();
        q.offer(new Point(0, 0));
        cost[0][0] = 1;
        
        int[] dx = {0, 0, -1, 1};
        int[] dy = {-1, 1, 0, 0};
        
        while (!q.isEmpty()) {
            Point now = q.poll();
            
            for (int i = 0; i < 4; i++) {
                int nx = dx[i] + now.x;
                int ny = dy[i] + now.y;
                
                if (0 <= nx && nx < col && 0 <= ny && ny < row) {
                    if (maps[nx][ny] == 1 && cost[nx][ny] > cost[now.x][now.y] + 1) {
                        q.offer(new Point(nx, ny));
                        cost[nx][ny] = cost[now.x][now.y] + 1;
                    }
                }
            }
        }
        
        return cost[col - 1][row - 1] != Integer.MAX_VALUE ? cost[col - 1][row - 1] : -1;
    }
}